Bootstrap command refactoring: consolidate output modes (step 3)

Kobzol opened the pull request:
This PR is a continuation to https://github.com/rust-lang/rust/pull/126731. It consolidates the output modes of bootstrap (`Print` vs `CaptureAll` vs `CaptureStdout`) and simplifies the logic around error printing (now a command error is always printed if the failure is not ignored). It also ports even more usages of `Command` to `BootstrapCommand`, most notably the git helpers and many usages of the `output` function.

The last commit was added because the third commit made two variants of the `Tool` enum unused (no idea why, but it seems to have been a false positive that they were used before).

It can be reviewed now, but I would wait with merging until at least a few days after https://github.com/rust-lang/rust/pull/126731, just to catch any potential issues from that PR before we move further.

As a next step, I want to clean up the API of the command a little bit to make usage easier (currently it's a bit verbose), and then continue with the rest of the tasks from the tracking issue.

As always, best reviewed commit by commit.

Tracking issue: https://github.com/rust-lang/rust/issues/126819

r? @onur-ozkan

try-job: aarch64-apple

Kobzol committed:
Remove temporary `BootstrapCommand` trait impls

Kobzol committed:
Make `git` helper return `BootstrapCmd`

Kobzol committed:
Remove various usages of the `output` function

Kobzol committed:
Allow unused `Tool` variants

Kobzol committed:
Improve documentation

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9723951470/job/26839653830) [(plain)](https://github.com/rust-lang/rust/commit/463bc8d06ca619e6e65c3899fcd16dc1aaca1a54/checks/26839653830/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9724177353/job/26840119766) [(plain)](https://github.com/rust-lang/rust/commit/31c398322cea086d845c547fdab0e9aea5777812/checks/26840119766/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
```

</details>

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9724486601/job/26840746292) [(plain)](https://github.com/rust-lang/rust/commit/b46d0899e9dfbdb6dc8af177d43223e7c181bbe3/checks/26840746292/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
    Finished `release` profile [optimized] target(s) in 13.74s
##[endgroup]


failed to execute command: BootstrapCommand { command: "lldb" "--version", failure_behavior: Ignore, output_mode: CaptureAll }


Build completed unsuccessfully in 0:12:25
  local time: Sat Jun 29 13:38:48 UTC 2024
```

</details>

Kobzol committed:
Refactor command outcome handling

To handle the case of failing to start a `BootstrapCommand`.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9724658112/job/26841116097) [(plain)](https://github.com/rust-lang/rust/commit/740f4e0f32d16ddff1697877f817806dd6368a47/checks/26841116097/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.35s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
```

</details>

bors commented:
:umbrella: The latest upstream changes (presumably #127121) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9726252398/job/26844491607) [(plain)](https://github.com/rust-lang/rust/commit/b92344e49fdf3df9857b70ff4f6a085a1f489fb3/checks/26844491607/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
  Downloaded boml v0.3.1
   Compiling boml v0.3.1
   Compiling y v0.1.0 (/checkout/compiler/rustc_codegen_gcc/build_system)
    Finished `release` profile [optimized] target(s) in 3.45s
     Running `/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-codegen/x86_64-unknown-linux-gnu/release/y test --use-system-gcc --use-backend gcc --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc --release --mini-tests --std-tests`
Using system GCC
[BUILD] example
[AOT] mini_core_hello_world
/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-tools/cg_gcc/mini_core_hello_world
abc
---
.....................

failures:

---- core::builder::tests::alias_and_path_for_library stdout ----
thread 'core::builder::tests::alias_and_path_for_library' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0

---- core::builder::tests::check_missing_paths_for_x_test_tests stdout ----
thread 'core::builder::tests::check_missing_paths_for_x_test_tests' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::check_missing_paths_for_x_test_tests' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::defaults::build_cross_compile stdout ----
thread 'core::builder::tests::defaults::build_cross_compile' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::defaults::build_cross_compile' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::defaults::build_default stdout ----
thread 'core::builder::tests::defaults::build_default' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::defaults::build_default' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::defaults::build_stage_0 stdout ----
thread 'core::builder::tests::defaults::build_stage_0' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::defaults::build_stage_0' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::defaults::doc_default stdout ----
thread 'core::builder::tests::defaults::doc_default' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::defaults::doc_default' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::dist::build_all stdout ----
thread 'core::builder::tests::dist::build_all' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::dist::build_all' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::dist::build_with_empty_host stdout ----
thread 'core::builder::tests::dist::build_with_empty_host' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::dist::build_with_empty_host' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::dist::dist_baseline stdout ----
thread 'core::builder::tests::dist::dist_baseline' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::dist::dist_baseline' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::dist::dist_only_cross_host stdout ----
thread 'core::builder::tests::dist::dist_only_cross_host' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::dist::dist_only_cross_host' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::dist::dist_with_empty_host stdout ----
thread 'core::builder::tests::dist::dist_with_empty_host' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::dist::dist_with_empty_host' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::dist::dist_with_hosts stdout ----
thread 'core::builder::tests::dist::dist_with_hosts' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::dist::dist_with_hosts' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::dist::dist_with_same_targets_and_hosts stdout ----
thread 'core::builder::tests::dist::dist_with_same_targets_and_hosts' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::dist::dist_with_same_targets_and_hosts' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::dist::dist_with_targets stdout ----
thread 'core::builder::tests::dist::dist_with_targets' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::dist::dist_with_targets' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::dist::dist_with_targets_and_hosts stdout ----
thread 'core::builder::tests::dist::dist_with_targets_and_hosts' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::dist::dist_with_targets_and_hosts' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::dist::doc_ci stdout ----
thread 'core::builder::tests::dist::doc_ci' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::dist::doc_ci' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::dist::llvm_out_behaviour stdout ----
thread 'core::builder::tests::dist::llvm_out_behaviour' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::dist::llvm_out_behaviour' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::dist::test_docs stdout ----
thread 'core::builder::tests::dist::test_docs' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::dist::test_docs' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::dist::test_with_no_doc_stage0 stdout ----
thread 'core::builder::tests::dist::test_with_no_doc_stage0' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::dist::test_with_no_doc_stage0' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::test_exclude stdout ----
thread 'core::builder::tests::test_exclude' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::test_exclude' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::test_exclude_kind stdout ----
thread 'core::builder::tests::test_exclude_kind' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::test_exclude_kind' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::test_valid stdout ----
thread 'core::builder::tests::test_valid' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::test_valid' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0
---- core::builder::tests::validate_path_remap stdout ----
thread 'core::builder::tests::validate_path_remap' panicked at src/core/metadata.rs:86:39:
thread 'core::builder::tests::validate_path_remap' panicked at src/core/metadata.rs:86:39:
serde_json::from_str(&metadata_output) failed with EOF while parsing a value at line 1 column 0

failures:
    core::builder::tests::alias_and_path_for_library
    core::builder::tests::check_missing_paths_for_x_test_tests
```

</details>

Kobzol committed:
Handle execution of dry run commands

bors commented:
:umbrella: The latest upstream changes (presumably #127133) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9732046814/job/26857236204) [(plain)](https://github.com/rust-lang/rust/commit/e6319499a52bccfa9fe828f411c410990a0f8e48/checks/26857236204/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Getting action download info
Download action repository 'msys2/setup-msys2@v2.22.0' (SHA:cc11e9188b693c2b100158c3322424c4cc1dadea)
Download action repository 'actions/checkout@v4' (SHA:692973e3d937129bcbf40652eb9f2f61becf3332)
Download action repository 'actions/upload-artifact@v4' (SHA:65462800fd760344b1a7b4382951275a0abb4808)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
---
COPY scripts/sccache.sh /scripts/
RUN sh /scripts/sccache.sh

COPY host-x86_64/mingw-check/reuse-requirements.txt /tmp/
RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt \
    && pip3 install virtualenv
COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#6 [internal] load build context
#6 transferring context: 7.11kB done
#6 DONE 0.0s

#9 [6/8] RUN pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv

#10 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#10 CACHED

```

</details>

Kobzol commented on src/bootstrap/src/utils/exec.rs:
```diff
@@ -32,17 +35,20 @@ pub enum OutputMode {
 /// If you want to delay failures until the end of bootstrap, use [delay_failure].
 ///
 /// By default, the command will print its stdout/stderr to stdout/stderr of bootstrap
```
Huh, I thought that doc comments need to be within 100 characters on a line, but tidy doesn't complain. Ok.

onur-ozkan replied:
Yeah tidy doesn't check line lengths in source files as it can be quite annoying to wrap lines that contain code.

Kobzol commented on src/bootstrap/src/utils/exec.rs:
```diff
@@ -13,16 +13,19 @@ pub enum BehaviorOnFailure {
+    /// Captures the stdout and stderr of the command into memory.
+    /// Corresponds to calling `cmd.output()`.
+    CaptureAll,
+    /// Captures the stdout of the command into memory, inherits stderr.
+    /// Corresponds to calling `cmd.output()`.
+    CaptureStdout,
```
Well, if you capture (only) stdout, you need to do something with stderr (instead of capturing it). You can either ignore it or inherit it (so that it will be printed to the stderr of bootstrap). Ignoring output seems bad, and all previous usages of this mode inherited stderr, so I suppose that it makes sense as a default?

- `CaptureStdout` - capture stdout, print stderr
- `CaptureAll` - capture both stdout and stderr

onur-ozkan replied:
> so I suppose that it makes sense as a default?

I have no objections to making it the default behaviour. My point was it's unclear to me that a type explicitly mentions `stdout` and inherits `stderr`, which I'm pretty sure most people will not expect until they read the type's doc-comment. It's just about the name-description clarity, not about whether this behavior should be the default or not.

Kobzol replied:
I see :) Any suggestions? CaptureStdoutPrintStderr?

onur-ozkan replied:
Yeah, that seems more clear.

Kobzol replied:
Tried to refactor this in a more general way, as the name reminded me of Java too much.

Kobzol commented on src/bootstrap/src/core/build_steps/tool.rs:
```diff
@@ -246,6 +244,7 @@ macro_rules! bootstrap_tool {
         ;
     )+) => {
         #[derive(PartialEq, Eq, Clone)]
+        #[allow(dead_code)]
```
Specifically it is `Tool::JsonDocCk` and `Tool::JsonDocLint`. For some reason, the compiler thought that they were used before this PR, probably because of some false positive (?). I haven't found anything in this PR that would remove their usage.

In general, I don't think that this should be a FIXME. The `Tool` enum is generated by a macro, and not all tools need to be built using `tool_cmd`, so it may be unused. The macro uses these two enum variants in the `tool_exe` function, but it is probably ignored by the compiler since the usage is inside a macro.

Kobzol committed:
Review changes

Kobzol commented:
@rustbot ready

Kobzol committed:
Simplify and generalize implementation of output mode

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/9783484214/job/27012222641) [(plain)](https://github.com/rust-lang/rust/commit/8dc4485109da76080a6b47cdd693d67431d0b9d0/checks/27012222641/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 46)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 49)
 - LegacyKeyValueFormat: "ENV key=value" should be used instead of legacy "ENV key value" format (line 61)
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
##[error]   --> src/utils/tarball.rs:373:17
    |
372 |             let timestamp = helpers::output(
    |                             --------------- arguments to this function are incorrect
373 |                 helpers::git(Some(&self.builder.src)).arg("log").arg("-1").arg("--format=%ct"),
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&mut Command`, found `&mut BootstrapCommand`
    = note: expected mutable reference `&mut std::process::Command`
               found mutable reference `&mut BootstrapCommand`
note: function defined here
   --> src/utils/helpers.rs:276:8
   --> src/utils/helpers.rs:276:8
    |
276 | pub fn output(cmd: &mut Command) -> String {

For more information about this error, try `rustc --explain E0308`.
error: could not compile `bootstrap` (lib) due to 1 previous error
failed to run: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml --locked
---
##[error]   --> src/utils/tarball.rs:373:17
    |
372 |             let timestamp = helpers::output(
    |                             --------------- arguments to this function are incorrect
373 |                 helpers::git(Some(&self.builder.src)).arg("log").arg("-1").arg("--format=%ct"),
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&mut Command`, found `&mut BootstrapCommand`
    = note: expected mutable reference `&mut std::process::Command`
               found mutable reference `&mut BootstrapCommand`
note: function defined here
   --> src/utils/helpers.rs:276:8
   --> src/utils/helpers.rs:276:8
    |
276 | pub fn output(cmd: &mut Command) -> String {

For more information about this error, try `rustc --explain E0308`.
error: could not compile `bootstrap` (lib) due to 1 previous error
failed to run: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml --locked
---
##[error]   --> src/utils/tarball.rs:373:17
    |
372 |             let timestamp = helpers::output(
    |                             --------------- arguments to this function are incorrect
373 |                 helpers::git(Some(&self.builder.src)).arg("log").arg("-1").arg("--format=%ct"),
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&mut Command`, found `&mut BootstrapCommand`
    = note: expected mutable reference `&mut std::process::Command`
               found mutable reference `&mut BootstrapCommand`
note: function defined here
   --> src/utils/helpers.rs:276:8
   --> src/utils/helpers.rs:276:8
    |
276 | pub fn output(cmd: &mut Command) -> String {

For more information about this error, try `rustc --explain E0308`.
error: could not compile `bootstrap` (lib) due to 1 previous error
failed to run: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml --locked
---
##[error]   --> src/utils/tarball.rs:373:17
    |
372 |             let timestamp = helpers::output(
    |                             --------------- arguments to this function are incorrect
373 |                 helpers::git(Some(&self.builder.src)).arg("log").arg("-1").arg("--format=%ct"),
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&mut Command`, found `&mut BootstrapCommand`
    = note: expected mutable reference `&mut std::process::Command`
               found mutable reference `&mut BootstrapCommand`
note: function defined here
   --> src/utils/helpers.rs:276:8
   --> src/utils/helpers.rs:276:8
    |
276 | pub fn output(cmd: &mut Command) -> String {

For more information about this error, try `rustc --explain E0308`.
error: could not compile `bootstrap` (lib) due to 1 previous error
failed to run: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml --locked
---
##[error]   --> src/utils/tarball.rs:373:17
    |
372 |             let timestamp = helpers::output(
    |                             --------------- arguments to this function are incorrect
373 |                 helpers::git(Some(&self.builder.src)).arg("log").arg("-1").arg("--format=%ct"),
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `&mut Command`, found `&mut BootstrapCommand`
    = note: expected mutable reference `&mut std::process::Command`
               found mutable reference `&mut BootstrapCommand`
note: function defined here
   --> src/utils/helpers.rs:276:8
   --> src/utils/helpers.rs:276:8
    |
276 | pub fn output(cmd: &mut Command) -> String {

For more information about this error, try `rustc --explain E0308`.
error: could not compile `bootstrap` (lib) due to 1 previous error
failed to run: /checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo build --manifest-path /checkout/src/bootstrap/Cargo.toml --locked
```

</details>

Kobzol committed:
Rebase on master

onur-ozkan commented:
@bors r+

bors commented:
:pushpin: Commit 54952b444566956b12058162eae383c692da0147 has been approved by `onur-ozkan`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=onur-ozkan 54952b444566956b12058162eae383c692da0147 -->
<!-- homu: {"type":"Approved","sha":"54952b444566956b12058162eae383c692da0147","approver":"onur-ozkan","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 54952b444566956b12058162eae383c692da0147 with merge f43964601a774724e2c556b56296a70bf5388efa...
<!-- homu: {"type":"BuildStarted","head_sha":"54952b444566956b12058162eae383c692da0147","merge_sha":"f43964601a774724e2c556b56296a70bf5388efa"} -->

rust-log-analyzer commented:
The job **`aarch64-apple`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/actions/runs/9787197722/job/27023326495) [(plain)](https://github.com/rust-lang-ci/rust/commit/f43964601a774724e2c556b56296a70bf5388efa/checks/27023326495/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9787197722/job/27023416891)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/actions/runs/9787197722/job/27023416891","builder_name":"checks-actions"} -->

Kobzol committed:
Do not consider LLDB version to be valid if it is empty

When dry run is enabled, the command for finding LLDB version would succeed, but return an empty string. This was inadvertently enabling a code path that should only be executed when the LLDB is actually present and its version is valid. This commit makes sure that if the version is empty, LLDB will be considered not found.

Kobzol commented:
@bors try

bors commented:
:hourglass: Trying commit c198c0e6d0556cec74d10aec998f4e229137c701 with merge e15709dec2e7237899641da79d49bbf6e185cdf2...
<!-- homu: {"type":"TryBuildStarted","head_sha":"c198c0e6d0556cec74d10aec998f4e229137c701","merge_sha":"e15709dec2e7237899641da79d49bbf6e185cdf2"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9790798521/job/27036430114)
Build commit: e15709dec2e7237899641da79d49bbf6e185cdf2 (`e15709dec2e7237899641da79d49bbf6e185cdf2`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9790798521/job/27036430114"},"merge_sha":"e15709dec2e7237899641da79d49bbf6e185cdf2"} -->

Kobzol commented:
Fixed problem with dry run in Apple job.

@bors r=onur-ozkan

bors commented:
:pushpin: Commit c198c0e6d0556cec74d10aec998f4e229137c701 has been approved by `onur-ozkan`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=onur-ozkan c198c0e6d0556cec74d10aec998f4e229137c701 -->
<!-- homu: {"type":"Approved","sha":"c198c0e6d0556cec74d10aec998f4e229137c701","approver":"onur-ozkan","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit c198c0e6d0556cec74d10aec998f4e229137c701 with merge e2cf31a6148725bde4ea48acf1e4fe72675257a2...
<!-- homu: {"type":"BuildStarted","head_sha":"c198c0e6d0556cec74d10aec998f4e229137c701","merge_sha":"e2cf31a6148725bde4ea48acf1e4fe72675257a2"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9793085000/job/27046099675)
Approved by: onur-ozkan
Pushing e2cf31a6148725bde4ea48acf1e4fe72675257a2 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"onur-ozkan","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9793085000/job/27046099675"},"merge_sha":"e2cf31a6148725bde4ea48acf1e4fe72675257a2"} -->

rust-timer commented:
Finished benchmarking commit (e2cf31a6148725bde4ea48acf1e4fe72675257a2): [comparison URL](https://perf.rust-lang.org/compare.html?start=9ffe52e05b063f5fea020d501216a35559ca8620&end=e2cf31a6148725bde4ea48acf1e4fe72675257a2&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=9ffe52e05b063f5fea020d501216a35559ca8620&end=e2cf31a6148725bde4ea48acf1e4fe72675257a2&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=9ffe52e05b063f5fea020d501216a35559ca8620&end=e2cf31a6148725bde4ea48acf1e4fe72675257a2&stat=max-rss)
<details>
<summary>Results (secondary 0.2%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 1.1%  | [0.8%, 1.7%]   | 4     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.7% | [-2.3%, -1.1%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=9ffe52e05b063f5fea020d501216a35559ca8620&end=e2cf31a6148725bde4ea48acf1e4fe72675257a2&stat=cycles:u)
<details>
<summary>Results (secondary -1.7%)</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -1.7% | [-2.6%, -0.8%] | 2     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=9ffe52e05b063f5fea020d501216a35559ca8620&end=e2cf31a6148725bde4ea48acf1e4fe72675257a2&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 718.904s -> 734.204s (2.13%)
**Artifact size**: 328.18 MiB -> 328.34 MiB (0.05%)
